core/src/main/java/jeeves/server/context/ServiceContext.java

```diff
@@ -52,21 +53,179 @@
+     *     <li>trace: check behavior and log unusal use</li>
+     *     <li>strict: raise illegal state exception for unusual behavior</li>
+     * </ul>
+     */
+    private static final ThreadLocalPolicy POLICY;
+    static {
+        String property = System.getProperty("jeeves.server.context.policy", "TRACE");
+        ThreadLocalPolicy policy;
+        try {
+            policy = ThreadLocalPolicy.valueOf(property.toUpperCase());
+        }
+        catch (IllegalArgumentException defaultToDirect) {
+            policy = ThreadLocalPolicy.DIRECT;
+        }
+        POLICY = policy;
+    }
+
+    /**
+     * Be careful with thread local to avoid leaking resources, set POLICY above to trace allocation.
+     */
     private static final InheritableThreadLocal<ServiceContext> THREAD_LOCAL_INSTANCE = new InheritableThreadLocal<ServiceContext>();
-    private UserSession _userSession = new UserSession();
-    private InputMethod _input;
-    private OutputMethod _output;
-    private Map<String, String> _headers;
-    private String _language;
-    private String _service;
-    private String _ipAddress;
-    private int _maxUploadSize;
-    private JeevesServlet _servlet;
-    private boolean _startupError = false;
-    private Map<String, String> _startupErrors;
+
+    /**
+     * Simple data structure recording service details.
+     *
+     * Lightweight data structure used logging service details such as name.
+     */
+    public static class ServiceDetails {
+        private String ipAddress;
+        private String service;
+        private String language;
+        public ServiceDetails(ServiceContext context){
+            this.ipAddress = context.getIpAddress();
+        }
+
+        public String getIpAddress() {
+            return ipAddress;
+        }
+
+        public String getService() {
+            return service;
+        }
+
+        public String getLanguage() {
+            return language;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ServiceDetails that = (ServiceDetails) o;
+            return Objects.equals(ipAddress, that.ipAddress) && service.equals(that.service) && Objects.equals(language, that.language);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(ipAddress, service, language);
+        }
+    }
+
+    /**
+     * Shared service context offering limited functionality.
+     *
+     * Use of service context has been assumed in many parts of the codebase. This shared "AppHandler" service context
+     * is used during Jeeves startup and has some protection from being  cleared. Additional managers such HarvestManager
+     * also use a shared service context to support background processes.
+     */
+    public static class AppHandlerServiceContext extends ServiceContext {
```

Comment:
suggestion only: Public static class could be broken out as its own top-level class, similar to BasicContext.